core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java

```diff
@@ -66,6 +103,138 @@ public void checkRead(Object key) {
+   }
+
+   @Override
+   public Collection<GlobalTransaction> getPartialTransactions() {
+      return Collections.unmodifiableCollection(partialTransactions.keySet());
+   }
+
+   @Override
+   public boolean canRollbackTransactionAfterOriginatorLeave(GlobalTransaction globalTransaction) {
+      boolean canRollback = availabilityMode == AvailabilityMode.AVAILABLE &&
+            !getLastStableTopology().getActualMembers().contains(globalTransaction.getAddress());
+      if (trace) {
+         log.tracef("Can rollback transaction? %s", canRollback);
+      }
+      return canRollback;
+   }
+
+   @Override
+   public void onTopologyUpdate(CacheTopology cacheTopology) {
+      boolean isStable = isTopologyStable(cacheTopology);
+      if (trace) {
+         log.tracef("On stable topology update. Has pending tx? %b. Is stable? %b. topology=%s",
+                    !partialTransactions.isEmpty(), isStable, cacheTopology);
+      }
+      if (isStable) {
+         if (!partialTransactions.isEmpty()) {
+            for (TransactionInfo transactionInfo : partialTransactions.values()) {
+               completeTransaction(transactionInfo, cacheTopology);
+            }
+         }
+      }
+   }
+
+   private void completeTransaction(final TransactionInfo transactionInfo, CacheTopology cacheTopology) {
+      final Collection<Address> affectedNodes = transactionInfo.getCommitNodes(cacheTopology);
+      NotifyingNotifiableFuture<Map<Address, Response>> notifyingFuture = new NotifyingFutureImpl<>();
+      notifyingFuture.attachListener(new FutureListener<Map<Address, Response>>() {
+         @Override
+         public void futureDone(Future<Map<Address, Response>> future) {
+            final GlobalTransaction globalTransaction = transactionInfo.getGlobalTransaction();
+            if (trace) {
+               log.tracef("Future done for transaction %s", globalTransaction);
+            }
+            try {
+               future.get();
+            } catch (Throwable e) {
+               if (trace) {
+                  log.tracef(e, "Exception for transaction %s", globalTransaction);
+               }
+               return;
+            }
+            if (trace) {
+               log.tracef("Performing cleanup for transaction %s", globalTransaction);
+            }
+            lockManager.unlock(transactionInfo.getLockedKeys(), globalTransaction);
+            partialTransactions.remove(globalTransaction);
+            TxCompletionNotificationCommand command = commandsFactory.buildTxCompletionNotificationCommand(null, globalTransaction);
+            //a little bit overkill, but the state transfer can happen during a merge and some nodes can receive the
+            // transaction that aren't in the original affected nodes.
+            //no side effects.
+            rpcManager.invokeRemotely(null, command, rpcManager.getDefaultRpcOptions(false));
```

Comment:
I guess StateProviderImpl could check for the transaction in the PartitionHandlingImpl before sending it to new nodes, but I agree it's safest to broadcast the command to everyone.